aws-lc-rs/src/cipher/key.rs

```diff
@@ -0,0 +1,200 @@
+use core::ptr::copy_nonoverlapping;
+use std::mem::{size_of, transmute, MaybeUninit};
+use std::os::raw::c_uint;
+use std::ptr;
+use zeroize::Zeroize;
+
+pub(crate) enum SymmetricCipherKey {
+    Aes128 {
+        raw_key: Aes128Key,
+        enc_key: AES_KEY,
+        dec_key: AES_KEY,
+    },
+    Aes256 {
+        raw_key: Aes256Key,
+        enc_key: AES_KEY,
+        dec_key: AES_KEY,
+    },
+    ChaCha20 {
+        raw_key: ChaCha20Key,
+    },
+}
+
+unsafe impl Send for SymmetricCipherKey {}
+
+// The AES_KEY value is only used as a `*const AES_KEY` in calls to `AES_encrypt`.
+unsafe impl Sync for SymmetricCipherKey {}
+
+impl Drop for SymmetricCipherKey {
+    fn drop(&mut self) {
+        // Aes128Key, Aes256Key and ChaCha20Key implement Drop separately.
+        match self {
+            SymmetricCipherKey::Aes128 {
+                enc_key, dec_key, ..
+            }
+            | SymmetricCipherKey::Aes256 {
+                enc_key, dec_key, ..
+            } => unsafe {
+                #[allow(clippy::transmute_ptr_to_ptr)]
+                let enc_bytes: &mut [u8; size_of::<AES_KEY>()] = transmute(enc_key);
+                enc_bytes.zeroize();
+                #[allow(clippy::transmute_ptr_to_ptr)]
+                let dec_bytes: &mut [u8; size_of::<AES_KEY>()] = transmute(dec_key);
+                dec_bytes.zeroize();
+            },
+            SymmetricCipherKey::ChaCha20 { .. } => {}
+        }
+    }
+}
+
+impl SymmetricCipherKey {
+    pub(crate) fn aes128(key_bytes: &[u8]) -> Result<Self, Unspecified> {
+        if key_bytes.len() != AES_128_KEY_LEN {
+            return Err(Unspecified);
+        }
+
+        unsafe {
+            let mut enc_key = MaybeUninit::<AES_KEY>::uninit();
+            #[allow(clippy::cast_possible_truncation)]
+            if 0 != AES_set_encrypt_key(
+                key_bytes.as_ptr(),
+                (key_bytes.len() * 8) as c_uint,
+                enc_key.as_mut_ptr(),
+            ) {
+                return Err(Unspecified);
+            }
+            let enc_key = enc_key.assume_init();
+
+            let mut dec_key = MaybeUninit::<AES_KEY>::uninit();
+            #[allow(clippy::cast_possible_truncation)]
+            if 0 != AES_set_decrypt_key(
+                key_bytes.as_ptr(),
+                (key_bytes.len() * 8) as c_uint,
+                dec_key.as_mut_ptr(),
+            ) {
+                return Err(Unspecified);
+            }
+            let dec_key = dec_key.assume_init();
+
+            let mut kb = MaybeUninit::<[u8; 16]>::uninit();
+            ptr::copy_nonoverlapping(key_bytes.as_ptr(), kb.as_mut_ptr().cast(), 16);
+            Ok(SymmetricCipherKey::Aes128 {
+                raw_key: Aes128Key(kb.assume_init()),
+                enc_key,
+                dec_key,
+            })
+        }
+    }
+
+    pub(crate) fn aes256(key_bytes: &[u8]) -> Result<Self, Unspecified> {
+        if key_bytes.len() != 32 {
```

Comment:
```suggestion
        if key_bytes.len() != AES_256_KEY_LEN {
```
Maybe we can depend on `AES_128_KEY_LEN` and `AES_256_KEY_LEN` for all the instances of `16` and `32` in this file (except for `chacha20`).